crates/ra_hir_expand/src/builtin_macro.rs

```diff
@@ -266,6 +272,40 @@ fn concat_expand(tt: &tt::Subtree) -> Result<(tt::Subtree, FragmentKind), mbe::E
     Ok((quote!(#text), FragmentKind::Expr))
 }
 
+fn relative_file(db: &dyn AstDatabase, call_id: MacroCallId, path: &str) -> Option<FileId> {
+    let call_site = call_id.as_file().original_file(db);
+    let path = RelativePath::new(&path);
+
+    db.resolve_relative_path(call_site, &path)
+}
+
+fn include_expand(
+    db: &dyn AstDatabase,
+    arg_id: EagerMacroId,
+    tt: &tt::Subtree,
+) -> Result<(tt::Subtree, FragmentKind), mbe::ExpandError> {
+    let path = tt
+        .token_trees
+        .get(0)
+        .and_then(|tt| match tt {
+            tt::TokenTree::Leaf(tt::Leaf::Literal(it)) => unquote_str(&it),
+            _ => None,
+        })
+        .ok_or_else(|| mbe::ExpandError::ConversionError)?;
+
+    let file_id =
+        relative_file(db, arg_id.into(), &path).ok_or_else(|| mbe::ExpandError::ConversionError)?;
+
+    // FIXME:
+    // Handle include as expression
+    let node =
+        db.parse_or_expand(file_id.into()).ok_or_else(|| mbe::ExpandError::ConversionError)?;
```

Comment:
Yeah, ideally, I think we should do something like `parse_to_token_tree(db.file_text(file_id))`. That is, it seems like we don't need to construct a syntax tree at this point at all.

Reply:
Yeah. It will be my next PR :)

Reply:
Submitted #3518